File: Neighbor.cpp
#include "Neighbor.h"

const double max_weight = 5000000;

Neighbor::Neighbor(int arg_target, double arg_weight)
    : target(arg_target), weight(arg_weight) {
  visible = false;
  loc.x = loc.y = 0;
  id = "00";
}

Neighbor::Neighbor() {
  target = 0;
  weight = 0;
  id = "00";
  loc.x = loc.y = 0;
  visible = false;
}

void Neighbor::setLocation(int x, int y) {
  loc.x = x;
  loc.y = y;
}

void Neighbor::setID(std::string a) { id = a; }

void Neighbor::setVisibility(bool vis) { visible = vis; }

bool Neighbor::getVisibility() { return visible; }

std::string Neighbor::getID() { return id; }

Point Neighbor::getLocation() { return loc; }

void Neighbor::DijkstraComputePaths(int source,
                                    const adjacency_list_t &adjacency_list,
                                    vector<double> &min_distance,
                                    vector<int> &previous) {
  int n = adjacency_list.size();
  min_distance.clear();
  min_distance.resize(n, max_weight);
  min_distance[source] = 0;
  previous.clear();
  previous.resize(n, -1);
  set<pair<double, int>> vertex_queue;
  vertex_queue.insert(make_pair(min_distance[source], source));

  while (!vertex_queue.empty()) {
    double dist = vertex_queue.begin()->first;
    int u = vertex_queue.begin()->second;
    vertex_queue.erase(vertex_queue.begin());

    // Visit each edge exiting u
    const vector<Neighbor> &neighbors = adjacency_list[u];
    for (vector<Neighbor>::const_iterator neighbor_iter = neighbors.begin();
         neighbor_iter != neighbors.end(); neighbor_iter++) {
      int v = neighbor_iter->target;
      double weight = neighbor_iter->weight;
      double distance_through_u = dist + weight;
      if (distance_through_u < min_distance[v]) {
        vertex_queue.erase(make_pair(min_distance[v], v));

        min_distance[v] = distance_through_u;
        previous[v] = u;
        vertex_queue.insert(make_pair(min_distance[v], v));
      }
    }
  }
}

list<int> Neighbor::DijkstraGetShortestPathTo(int vertex,
                                              const vector<int> &previous) {
  list<int> path;
  for (; vertex != -1; vertex = previous[vertex])
    path.push_front(vertex);
  return path;
}

bool Neighbor::onNode(int x, int y, int ShiftFactorX, int ShiftFactorY) {
  Point a;
  a.x = x - ShiftFactorX;
  a.y = y - ShiftFactorY;
  double radius = 12;
  double dist = a.distance(loc);
  return dist < radius;
}

void Neighbor::setNodeIsPressed(bool newVal) { NodeIsPressed = newVal; }

void Neighbor::setoverNode(bool newVal) { overNode = newVal; }

bool Neighbor::getNodeIsPressed() { return NodeIsPressed; }

File: python/neighbor.py
from point import distance

class Neighbor(object):
  max_weight = 5000000

  def __init__(self, target=0, weight=0):
    self.target = target
    self.weight = weight
    self.visible = False
    self.over_node = False
    self.node_is_pressed = False
    self.location = (0, 0)
    self.id = '00'


  def dijkstra_compute_paths(self, source, adjacency_list):
    n = len(adjacency_list)

    min_distance = [Neighbor.max_weight for _ in range(n)]
    min_distance[source] = 0

    previous = [-1 for _ in range(n)]

    vertex_queue = set()
    vertex_queue.add((min_distance[source], source))

    while vertex_queue:
      dist, u = vertex_queue.pop()
      vertex_queue.remove((dist, u))

      # Visit each edge exiting u
      neighbors = adjacency_list[u]
      for neighbor in neighbors:
        v = neighbor_iter.target
        weight = neighbor.weight;
        distance_through_u = dist + weight
        if distance_through_u < min_distance[v]:
          vertex_queue.discard((min_distance[v], v))

          min_distance[v] = distance_through_u
          previous[v] = u
          vertex_queue.add((min_distance[v], v))

    return min_distance, previous


  def dijkstra_get_shortest_path_to(self, vertex, previous):
    path = []
    while vertex != -1:
      path.append(vertex)
      vertex = previous[vertex]
    return reversed(path)


  def on_node(self, x, y, shift_factor_x, shift_factor_y):
    radius = 12
    dist = distance((x - shift_factor_x, y - shift_factor_y), self.location)
    return dist < radius
